### PR TITLE
Metrics absolute average fold error

### PR DESCRIPTION
**Pull request by zhu0619**
## Changelogs

- Add metric Balanced average precision from sklearn 
- Add metric Absolute Average Fold Error implementation from https://pubs.acs.org/doi/10.1021/acs.chemrestox.3c00305

---

_Checklist:_

- [x] _Add tests to cover the fixed bug(s) or the newly introduced feature(s) (if appropriate)._
- [x] _Update the API documentation if a new function is added, or an existing one is deleted._
- [x] _Write concise and explanatory changelogs above._
- [x] _If possible, assign one of the following labels to the PR: `feature`, `fix` or `test` (or ask a maintainer to do it for you)._

---

_discussion related to that PR_

### CODE REVIEW

File: polaris/evaluate/_metric.py [hunk 1]
```diff
@@ -30,6 +31,28 @@ def spearman(y_true: np.ndarray, y_pred: np.ndarray):
...
+
+    Parameters:
+    y_true : array-like of shape (n_samples,)
+        The true target values.
+    y_pred : array-like of shape (n_samples,)
+        The predicted target values.
+
+    Returns:
+    aafe : float
+        The Absolute Average Fold Error.
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:46]: Should be consistent with the docstring style in the rest of the repo, so `Args` instead of `Parameters` and no type annotations in the docstring.

File: polaris/evaluate/_metric.py [hunk 2]
```diff
@@ -30,6 +31,28 @@ def spearman(y_true: np.ndarray, y_pred: np.ndarray):
     return stats.spearmanr(y_true, y_pred).statistic
 
 
+def absolute_average_fold_error(y_true: np.ndarray, y_pred: np.ndarray):
+    """
+    Calculate the Absolute Average Fold Error (AAFE) metric.
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:36]: Could you explain what this metric measures? A brief description + a citation would be enough!

File: polaris/evaluate/_metric.py [hunk 3]
```diff
@@ -75,7 +99,9 @@ class Metric(Enum):
     pr_auc = MetricInfo(fn=average_precision_score, direction="max")
     mcc = MetricInfo(fn=matthews_corrcoef, direction="max")
     cohen_kappa = MetricInfo(fn=cohen_kappa_score, direction="max")
+
     # TODO: adding metrics for multiclass tasks
+    ba = MetricInfo(fn=balanced_accuracy_score, direction="max")
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:104]: For clarity, I would use:
```suggestion
    balanced_accuracy = MetricInfo(fn=balanced_accuracy_score, direction="max")
```

File: polaris/evaluate/_metric.py [hunk 4]
```diff
@@ -65,6 +88,7 @@ class Metric(Enum):
     pearsonr = MetricInfo(fn=pearsonr, direction="max")
     spearmanr = MetricInfo(fn=spearman, direction="max")
     explained_var = MetricInfo(fn=explained_variance_score, direction="max")
+    aafe = MetricInfo(fn=absolute_average_fold_error, direction="max")
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:91]: For clarity, I would use:
```suggestion
    absolute_average_fold_error = MetricInfo(fn=absolute_average_fold_error, direction="max")
```

File: polaris/evaluate/_metric.py (hunk 3 above)
Comment (cwognum) [reply polaris/evaluate/_metric.py:104]: Also, by placing this under the `TODO` comment, it seems to me this is a multi-task metric? Is that the case? If so, the `MetricInfo.is_multitask` flag should be set and I would update the comment. If not, I would move it above the comment.

**Review by cwognum — CHANGES_REQUESTED** [review]
Thanks, @zhu0619 

For this to be merged, there needs to be a similar PR on the Hub side as well that adds support for these new metrics. Is this something you would feel comfortable doing, @zhu0619 ? If not, I can look into it!

**Comment by cwognum** [comment]
Ah I see you created https://github.com/polaris-hub/polaris-hub/issues/317.

File: polaris/evaluate/_metric.py (hunk 3 above)
Comment (zhu0619) [reply polaris/evaluate/_metric.py]: It's multiclass not multitask.

**Review by cwognum — CHANGES_REQUESTED** [review]
Could we add some test cases? Especially for the multi-class one, this sounds like something that could fail.

**Comment by zhu0619** [comment]
> Thanks, @zhu0619
> 
> For this to be merged, there needs to be a similar PR on the Hub side as well that adds support for these new metrics. Is this something you would feel comfortable doing, @zhu0619 ? If not, I can look into it!

I will do it.

File: polaris/evaluate/_metric.py [hunk 5]
```diff
@@ -92,18 +92,21 @@ class Metric(Enum):
...
     accuracy = MetricInfo(fn=accuracy_score, direction="max")
+    balanced_accuracy = MetricInfo(fn=balanced_accuracy_score, direction="max")
+    mcc = MetricInfo(fn=matthews_corrcoef, direction="max")
+    cohen_kappa = MetricInfo(fn=cohen_kappa_score, direction="max")
+
+    # binary only
     f1 = MetricInfo(fn=f1_score, kwargs={"average": "binary"}, direction="max")
-    f1_macro = MetricInfo(fn=f1_score, kwargs={"average": "macro"}, direction="max")
-    f1_micro = MetricInfo(fn=f1_score, kwargs={"average": "micro"}, direction="max")
+    # note: At the moment, multi-dimension inputs for classification are not supported
```
Comment (zhu0619) [inline polaris/evaluate/_metric.py:103]: @cwognum We can extend input attributes with `y_prob` so user can provide both predicted classes and probabilities for different multiclass metrics.

Comment (zhu0619) [reply polaris/evaluate/_metric.py:103]: It doesn't have to be addressed in this PR though.

File: polaris/evaluate/_metric.py [hunk 6]
```diff
@@ -30,6 +31,30 @@ def spearman(y_true: np.ndarray, y_pred: np.ndarray):
...
+
+    Args:
+        y_true : array-like of shape (n_samples,)
+            The true target values.
+        y_pred : array-like of shape (n_samples,)
+            The predicted target values.
+
+    Returns:
+        aafe : float
+            The Absolute Average Fold Error.
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:48]: We don't annotate the types in the docstring because we use type-hints already!  

```suggestion
    Args:
        y_true: The true target values of shape (n_samples,)
        y_pred: The predicted target values of shape (n_samples,).

    Returns:
        aafe: The Absolute Average Fold Error.
```

File: polaris/evaluate/_metric.py [hunk 7]
```diff
@@ -30,6 +31,30 @@ def spearman(y_true: np.ndarray, y_pred: np.ndarray):
     return stats.spearmanr(y_true, y_pred).statistic
 
 
+def absolute_average_fold_error(y_true: np.ndarray, y_pred: np.ndarray):
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:34]: This is actually needed, because the docs will fail otherwise. 

```suggestion
def absolute_average_fold_error(y_true: np.ndarray, y_pred: np.ndarray) -> float:
```

File: polaris/evaluate/_metric.py [hunk 8]
```diff
@@ -30,6 +31,30 @@ def spearman(y_true: np.ndarray, y_pred: np.ndarray):
...
+    """
+    Calculate the Absolute Average Fold Error (AAFE) metric.
+    It measures the fold change between predicted values and observed values.
+    The implementation is based on https://pubs.acs.org/doi/10.1021/acs.chemrestox.3c00305.
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:38]: Looks nice in the rendered docs! 

```suggestion
    The implementation is based on [this paper](https://pubs.acs.org/doi/10.1021/acs.chemrestox.3c00305).
```

File: polaris/evaluate/_metric.py (hunk 5 above)
Comment (cwognum) [reply polaris/evaluate/_metric.py:103]: That's a good point. We probably should and add a `needs_probs` attribute to the `MetricInfo` class?

File: tests/test_evaluate.py
Comment (cwognum) [inline tests/test_evaluate.py:1]: Could we also add a unit test to simply check if calling `absolute_average_fold_error` works as expected, since it's a custom implementation? Can be simple! Just some hard-coded values to compare against.

File: polaris/evaluate/_metric.py (hunk 7 above)
Comment (cwognum) [inline polaris/evaluate/_metric.py:34]: Let's add this to our API docs. You can add the following to `docs/api/evaluation.md`:

```

::: polaris.evaluate._metric.absolute_average_fold_error

---
```

**Review by cwognum — CHANGES_REQUESTED** [review]
Hey @zhu0619 , this is probably a bigger change than you anticipated by now! Sorry. Think we're almost there! Let me know if you need help. Happy to jump in.

File: polaris/evaluate/_metric.py [hunk 9]
```diff
@@ -65,17 +90,23 @@ class Metric(Enum):
     pearsonr = MetricInfo(fn=pearsonr, direction="max")
     spearmanr = MetricInfo(fn=spearman, direction="max")
     explained_var = MetricInfo(fn=explained_variance_score, direction="max")
+    absolute_average_fold_error = MetricInfo(fn=absolute_average_fold_error, direction="max")
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:93]: Based on our discussion in person, just adding this comment so we don't forget: We should check the `direction`!

Comment (cwognum) [reply polaris/evaluate/_metric.py:93]: I believe the target is `1.0`, with a range of `[0, ∞)`.

Comment (cwognum) [reply polaris/evaluate/_metric.py:93]: Actually, if targets are 0.0 there could be a division by zero. We should probably add a small delta to prevent that? Or check for it?

File: polaris/evaluate/_metric.py [hunk 10]
```diff
@@ -45,6 +70,7 @@ class MetricInfo(BaseModel):
     is_multitask: bool = False
     kwargs: dict = Field(default_factory=dict)
     direction: DirectionType
+    needs_probs: bool = False
```
Comment (cwognum) [inline polaris/evaluate/_metric.py:73]: In line with scikit-learn, I would like to use named arguments that are specific to what's expected, e.g. `def metric_1(y_true, y_pred)` vs `def metric_2(y_true, y_prob)`. I find that to be more clear and like to be consistent with Scikit-learn.

It seems scikit-learn uses `y_pred`, `y_prob` and `y_score`. What would you think of: 
```suggestion
    y_type: Literal["y_pred", "y_prob", "y_score"] = "y_pred"
```
?

 Then in `Metric.score()` we can do something like: 
```python
def score(
    self, 
    y_true: np.ndarray, 
    y_pred: Optional[np.ndarray] = None, 
    y_score: Optional[np.ndarray] = None, 
    y_prob: Optional[np.ndarray] = None
) -> float:
        
    if y_pred is None and y_score is None and y_prob is None: 
        raise ValueError(...)

    if self.y_type == "y_pred":
        pred = y_pred
    elif self.y_type == "y_score":
        pred = y_score
    else:
        pred = y_prob

    kwargs = {"y_true": y_true, self.y_type: pred}
    return self.fn(**kwargs)
```

I think it's more robust, but night complicate the API for the user, because you would probably have to propagate such changes to the API of the benchmark too.

What do you think?

File: tests/test_evaluate.py [hunk 11]
```diff
@@ -73,17 +73,33 @@ def test_metrics_singletask_clf(
...
+    result = test_single_task_benchmark_multi_clf.evaluate(predictions, probablities)
+    for metric in test_single_task_benchmark_multi_clf.metrics:
+        assert metric in result.results.Metric.tolist()
+
+
 def test_metrics_multitask_clf(tmpdir: str, test_multi_task_benchmark_clf: MultiTaskBenchmarkSpecification):
     train, test = test_multi_task_benchmark_clf.get_train_test_split()
     predictions = {
         target_col: np.random.randint(2, size=test.inputs.shape[0]) for target_col in train.target_cols
     }
```
Comment (cwognum) [inline tests/test_evaluate.py:98]: This can be removed, right?

**Review by cwognum — CHANGES_REQUESTED** [review]
Hey @zhu0619, I had one big question I would be curious to hear your thoughts on. Other than that, just some small comments.

File: polaris/benchmark/_base.py [hunk 12]
```diff
@@ -408,6 +408,8 @@ def evaluate(self, y_pred: PredictionsType) -> BenchmarkResults:
                 If there are multiple targets, the predictions should be wrapped in a dictionary with the target labels as keys.
                 If there are multiple test sets, the predictions should be further wrapped in a dictionary
                     with the test subset labels as keys.
+            y_prob: The predicted probabilities for the test set, as NumPy arrays.
+                Currently only multiclass in singletask setting is supported.
```
[the remark this answers is not in the document]
Comment (zhu0619) [reply polaris/benchmark/_base.py:412]: I don't think we have metrics which are able to handle multitask multiclass with mutli dimentsional probablities ?

File: polaris/evaluate/_metric.py (hunk 10 above)
Comment (zhu0619) [reply polaris/evaluate/_metric.py:73]: It would work. But isn't it additional work that user who defines metrics have to figure out the "y_type" for each metric?

Comment (zhu0619) [reply polaris/evaluate/_metric.py:73]: Maybe use `metric_args = inspect.signature(self.fn).parameters` to get the `y_type` instead.

Comment (zhu0619) [reply polaris/evaluate/_metric.py:73]: > It would work. But isn't it additional work that user who defines metrics have to figure out the "y_type" for each metric?

oh. I see. It's defined in the MetricInfo.

Comment (zhu0619) [reply polaris/evaluate/_metric.py:73]: The sklearn metrics used various argument names for `y_pred`. https://scikit-learn.org/stable/modules/model_evaluation.html#classification-metrics
Which is fine because they will be defined internally with MetricInfo. 
I am not sure whether to differenciate "y_score" and "y_prob". To me, one of them would be enough. 
In Benchmark.evaluate(), I think we can simplify them to either "y_score" or "y_prob". Maybe "y_score" is more universal.

File: polaris/benchmark/_base.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (cwognum) [reply polaris/benchmark/_base.py:412]: Ah okay, so it's a limitation of the metrics? If so, I would remove the comment or add a `TODO` to the `Metric` class. I thought it was a limitation of the `BenchmarkSpecification.evaluate()` method!

File: polaris/evaluate/_metric.py (hunk 10 above)
Comment (cwognum) [reply polaris/evaluate/_metric.py:73]: > In Benchmark.evaluate(), I think we can simplify them to either "y_score" or "y_prob".

Yes, agreed, but let's go with `y_prob` as you have it now.

From the scikit-learn docs on `y_score` (see also [here](https://scikit-learn.org/stable/glossary.html#term-decision_function)): 
> These can be either probability estimates or non-thresholded decision values

It seems that if we expect `y_score`, we can always provide `y_prob`, but not the other way around. I thus think `y_prob` will be the better choice and I also think its naming is better understood.

File: polaris/benchmark/_base.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (zhu0619) [reply polaris/benchmark/_base.py:412]: Yeah. It's the limitation of the metrics. I will add TODO.

File: tests/test_evaluate.py (hunk 11 above)
Comment (zhu0619) [reply tests/test_evaluate.py:98]: The evaluation will take both 'predictions' and 'probabilities' as input for all the defined metrics.

File: polaris/benchmark/_base.py [hunk 13]
```diff
@@ -443,13 +447,17 @@ def evaluate(self, y_pred: PredictionsType) -> BenchmarkResults:
             for metric in self.metrics:
                 if metric.is_multitask:
                     # Multi-task but with a metric across targets
-                    score = metric(y_true=y_true_subset, y_pred=y_pred[test_label])
+                    score = metric(
+                        y_true=y_true_subset, y_pred=y_pred[test_label], y_prob=y_prob.get(test_label, None)
```
[the remark this answers is not in the document]
Comment (cwognum) [reply polaris/benchmark/_base.py:451]: I am not sure if we're saying the same thing. I meant to say that `{}.get()` will return `None`. No need to specify the second (i.e. `default`) parameter.

**Review by cwognum — APPROVED** [review]
Thanks, @zhu0619 ! Looks great!
